### ISSUE TITLE
improve i/o error messages

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#699]
from https://github.com/google/xi-mac/issues/188. Error messages that may be displayed by the client should be reasonably descriptive.

In this case, errors that occur when trying to read a file should indicate the file in question and the error type (encoding, permissions, file not found etc).

**Comment by leomikezee** [comment]
I want to work on this issue.
Currently, I think all FileError are converted to RemoteError with ID:5, Message: "failed to load file", Data: None.
```
impl From<FileError> for RemoteError {
    fn from(_src: FileError) -> RemoteError {
        //TODO: when we migrate to using the failure crate for error handling,
        // this should return a better message
        RemoteError::custom(5, "failed to load file", None)
    }
}
```
However, we have 3 kinds of FileError, and io::Error may have more ErrorKinds itself, do we want to encode these errors with different id?
```
pub enum FileError {
    Io(io::Error),
    UnknownEncoding,
    HasChanged,
}
```
In addition, if we want to show the file name, we may need to add file path to the FileError enum and convert errors to FileError explicitly instead of using From trait. Something like this:
```
pub enum FileError {
    Io(io::Error, Path),
    UnknownEncoding(Path),
    HasChanged(Path),
}
```

**Comment by cmyr** [comment]
Error numbers: currently these are... haphazard. I'm not sure what their long-term status is going to be, but it makes sense to at least be a bit more rationale about them; set aside some range of numbers for I/O errors, and assign one per error type.

Augmenting the individual errors with whatever information they need to be useful also makes sense. This may be just the path, but I could also imagine other stuff where we have it.

Our current error handling story is bad, and there is a theoretical desire to improve it. In particular, we don't currently have any way of handling errors that happen during RPCs that do not already expect a result, and _most_ RPCs do not expect a result. There's a discussion issue for this: #543. 

I'm also curious about things like the [failure crate](https://crates.io/crates/failure), which is an attempt to improve error handling in Rust more generally.

That said, there is definitely room for improvement in this particular case regardless of whether we make those changes, but maybe this context will be helpful?

**Comment by leomikezee** [comment]
Made a PR #708 , and here is what it looks like now.
![image](https://user-images.githubusercontent.com/7655600/41393340-4299fc6e-6f73-11e8-985c-8a860947b4a7.png)

**Comment by leomikezee** [comment]
PR #708 approved

**Comment by cmyr** [comment]
@leomikezee not sure if you know this or not, but there are some nice tricks for associating issues with PRs, so you can do things like have an issue that is automatically closed when a PR is merged:

https://help.github.com/articles/closing-issues-using-keywords/

**Comment by leomikezee** [comment]
@cmyr Thanks, I will try it next time.

**Comment by cmyr** [comment]
Fixed in #708, thanks @leomikezee!
